platform/engineering/design-docs/separate-content-build.md

```diff
@@ -0,0 +1,373 @@
+
+Risks include:
+- Misconfiguring the forward proxy routing
+  - Could "remove" all static content from VA.gov
+  - Could "remove" all applications from VA.gov
+
+### Work Estimates
+The following estimates vary greatly depending on who's doing the work.
+
+1. Write a standalone script that can be pointed to the build output directory
+   to check for broken links
+    - Run this in Jenkins (Nomad? Circle?) after the build succeeds
+    - **Estimate:** 3 hours - 3 days
+1. Remove the broken link checker step from the Metalsmith script
+    - **Estimate:** < 1 hour
+1. Separate the content and application builds into their own scripts within
+   `vets-website`
+    - **Estimate:** 1 - 5 days
+1. Coordinate the builds with a consolidated build script to produce the same
+   output of the current build
+    - **Estimate:** 2 hours - 2 days
+1. Copy the content build script to another repo
+    - **Estimate:** 2 hours - 1 day
+1. Set up the CI for that new repo
+    - **Estimate:** ??
+    - I'm not sure what all goes into this
+        - What does it take to wire this into Jenkins / Nomad?
+1. Create landing pages within the CMS for each application
+    - **TO DO:** Ensure this won't override the existing landing pages yet
+    - **Estimate:** 1 hour - 1 day
+    - **Note:** This work can be done in parallel with any of the above tasks
+1. Coordinate with Ops to use the new repo to deploy content to a new bucket
+    - **Estimate:** ??
+    - Again, not sure what it takes to make this happen
+1. Set up the routing in the forward proxy
+    - **Estimate:** ??
+    - Probably between a few hours and a week...?
+1. Once we're confident all traffic is routed properly, switch the `vets-website`
+   build to build only webpack assets (content on the application bucket
+   shouldn't be touched at this point anyhow)
+    - **Estimate:** < 1 hour
+
+### Alternatives
+
+#### Mono-repo
+The alternative to splitting out the content build into its own separate repo is
+to keep it in `vets-website`. The reasons we're not doing this are:
+- We're trying to reduce what `vets-website` is "in charge" of
+  - Building static content from the CMS and compiling application code are two
+    completely separate functions
+- We can have one CI pipeline per repository
+  - And they'll be more narrowly focused
+- We can manage dependencies independently
+  - All dependencies in the content build repo are not client-facing
+    - There _may_ be exceptions later
+  - It will be harder for an application to use be able to use a dependency that
+    wasn't intended to be client-facing
+
+#### Single S3 bucket
+Instead of deploying content to a separate S3 bucket from the applications, we
+could continue to deploy them both to the same bucket. We decided to split up
+the deployment to:
+- Promote a higher separation of concerns
+  - **Question:** Is this even a thing we should be worried about?
+- Enable simpler rollbacks
+  - **Question:** Is this valid?
```

Comment:
Something to consider here is the 100 bucket limit per account. If we grow greater than 100 teams, this will be a problem:
https://docs.aws.amazon.com/AmazonS3/latest/dev/BucketRestrictions.html

Single bucket multiple folders is what I would suggest. s3 has a flat structure and it treats the folder structure as keys to the object. There is only a limit on the object size and not the bucket size:
https://docs.aws.amazon.com/AmazonS3/latest/user-guide/using-folders.html

Reply:
Ooh, interesting. :thinking:

Okay, so you're proposing achieving the same _kind_ of isolation of the applications and content by using separate directories in a single bucket?

I think the reason I came up with the multiple bucket idea was because I couldn't figure out how independent rollback would work, but with versioned folders like you mentioned in our chat the other day, I think that pretty much solves that problem.

I'd be down for this single-bucket approach myself, but I'd like to hear opinions from @nathanhruby or @wyattwalter.

Reply:
Would this require changing the URL in the `<script/>` tag on every release?

Reply:
the 100 bucket limit can be raised to 1000 whihc is not a limit I think we'd hit.  

Note that for static content sites the bucket must be named after the site being served, so with a single bucket you are also going to force a single hostname (as is the case now).